concourse/steps/release.py

```diff
@@ -23,19 +28,303 @@
 )
 
 
-def rebase(
-    git_helper,
-    upstream_ref,
-):
-    upstream_commit_sha = git_helper.fetch_head(upstream_ref).hexsha
-    git_helper.rebase(commit_ish=upstream_commit_sha)
+class ReleaseContext(object):
+    def set_release_commit_sha(self, release_commit_sha):
+        self._release_commit_sha = release_commit_sha
+
+    def release_commit_sha(self):
+        return self._release_commit_sha
+
+
+class TransactionalStep(object):
+    ''' Provides a framework to implement some action in an atomic fashion.
+
+    Instances of this class are intended to perform exactly one change in their apply-method
+    and provide the corresponding undo-functionality in the revert-method
+    '''
+    def __init__(self, release_context: ReleaseContext):
+        self.release_context = release_context
+
+    def validate(self):
+        pass
+
+    def apply(self):
+        raise NotImplementedError
+
+    def revert(self):
+        pass
+
+    def name(self):
+        pass
+
+
```

Comment:
I have already mentioned this in the documentation proposal. I suggest to add a class representing a sequence of steps. Maybe call it `Transaction`. It should receive the step instances and offer a method to run all steps in a transactional fashion (which you inlined a but further down).

That way (with the changes proposed up to here), we could consider to mv this out of this module and re-use it for other steps. This is however a change for another PR (the moving-out and re-use. the Transaction class should be introduced here).